### ISSUE TITLE
hmm bw seems wrong

### ISSUE THREAD

**Issue by alexrudnick** [nltk/nltk#100]
<b>What version of NLTK are you using?  (See nltk.**version**).  Please only</b>
<b>submit bug reports for the current version.</b>
2.0b7 

<b>What steps will reproduce the problem?  (e.g. include Python source code)</b>
iteration 0 logprob -1465.36185134
iteration 1 logprob 0
iteration 2 logprob 0
convergence too early
I think the problem is that update strategy is wrong.
&quot;beta[T-1, :] = log2(1)&quot; I think beta should take the highest alpha[T-1,:]
as 1 others as 0 if the end state is not restricted
When I debug that, I found model._transition does not update after the
first iterate

<b>What is the expected output? What do you see instead?</b>

<b>Please use labels and text to provide additional information.</b>

This is my email if someone wants to discuss:
angelhelee@gmail.com

<p>Migrated from http://code.google.com/p/nltk/issues/detail?id=513</p>

**Comment by tyleha** [comment]
I find this precise issue to still be current. No matter the input sequence, or the order in which it is input, or the tolerance for change in logprob between iters, all this algorithm does is update the ._output matrix to accurately represent the occurrence of each symbol in the sequence, and then quit. 

**Verson of NLTK**
2.0.3

**Here are some steps to reproduce the problem.**

s = """"Your humble writer knows a little bit about a lot of things, but despite writing a fair amount about text processing (a book, for example), linguistic processing is a relatively novel area for me. Forgive me if I stumble through my explanations of the quite remarkable Natural Language Toolkit (NLTK), a wonderful tool for teaching, and working in, computational linguistics using Python. Computational linguistics, moreover, is closely related to the fields of artificial intelligence, language/speech recognition, translation, and grammar checking.\nWhat NLTK includes\nIt is natural to think of NLTK as a stacked series of layers that build on each other. Readers familiar with lexing and parsing of artificial languages (like, say, Python) will not have too much of a leap to understand the similar -- but deeper -- layers involved in natural language modeling."""
sentences = s.split('.')[:-1]
seq = [map(lambda x:(x,''), ss.split(' ')) for ss in sentences]
symbols = list(set([ss[0] for sss in seq for ss in sss]))
states = range(5)

trainer = nltk.tag.hmm.HiddenMarkovModelTrainer(states=states,symbols=symbols)
m = trainer.train_unsupervised(seq)
m.random_sample(random.Random(),10)

**What is the expected output? What do you see instead?**
We would expect that m._priors._data and m._transitions would have been updated using the Baum-Welch algorithm. However, these two matrices remain identical to their initial distributions (that is to say log2(1/numofstates, and log2(1/numofsymbols), respectively). 

Example: 
iteration 0 logprob -2843.42261771
iteration 1 logprob -2661.17748749
iteration 2 logprob -2661.17748749

I too would be happy to help if someone with a better understanding of the Baum-Welch algorithm can give me a push in the right direction. tyleha@gmail.com is a place to reach me in that case.

**Comment by kmike** [comment]
Symptoms of the original report should be now fixed. 

`nltk.hmm.demo_pos_bw()` used to produce this:

```
>>> hmm.demo_pos_bw()

Baum-Welch demo for POS tagging

Training HMM (supervised)...
Training (unsupervised)...
iteration 0 logprob -1465.36185134
iteration 1 logprob 0
iteration 2 logprob 0
```

Numbers match the report so I assume the report is about the output of `nltk.hmm.demo_pos_bw()`.

After fixing an error in demo code it produces a more sensible result:

```
>>> hmm.demo_pos_bw()

Baum-Welch demo for POS tagging

Training HMM (supervised)...
Training (unsupervised)...
iteration 0 logprob -1465.36185134
iteration 1 logprob -1145.91490385
iteration 2 logprob -951.511474783
iteration 3 logprob -853.002994211
iteration 4 logprob -866.48166457
```

I didn't check if claims in original report are correct ("update strategy is wrong", "model._transition does not update after the first iterate"); I also didn't check a comment by @tylerhartley yet, so this ticket should probably remain open.

**Comment by kmike** [comment]
Update strategy in `HiddenMarkovModelTagger._backward_probability` looks correct; setting beta[T-1:] to 1 is what proposed in Rabiner tutorial; NLTK implementation uses log2 values so we have log2(1). 

I wrote some tests based on Wikipedia article on Forward-Backward algorithm, and results returned by `HiddenMarkovModelTagger._backward_probability` and `HiddenMarkovModelTagger._forward_probability` agree with calculations in Wikipedia; the results for `HiddenMarkovModelTagger._forward_probability` also agree with the example data in "Spoken Language Processing"  (Huang et al).

So if there is a bug it is likely in `HiddenMarkovModelTrainer.train_unsupervised`.

By the way,

```
iteration 3 logprob -853.002994211
iteration 4 logprob -866.48166457
```

looks suspicious because (from my understanding) logprob shouldn't ever decrease - it is proven that Baum-Welch transformation increases probability (could it be an issue with normalization? - alpha values are not normalized so sum(2**alpha[T-1, :]) != 1 in NLTK implementation).

**Comment by kmike** [comment]
@tylerhartley : priors are not updated in train_unsupervised by design (look at this comment: https://github.com/nltk/nltk/blob/fe3775f5f7e3c7d625d213e8b8091f8e7c5c71f5/nltk/tag/hmm.py#L981 )

I think that transitions are not updated because of the initialization; maybe using non-uniform initial distribution could help - but there of course can be a bug somewhere.

"model._transition does not update after the first iterate" was probably caused by one of the bugs fixed in #371; I can't reproduce this now.

I think this ticket may be closed when #371 is merged - BW would be still wrong, but for another reasons.

**Comment by tyleha** [comment]
@kmike - since adding to this ticket, I have independently confirmed using Matlab's hmm package that unifrorm flat distribution of `_priors` and `_transition` matrices results in exactly the behavior I reported (i.e. all elements in the matrix are of equal probability). Without any jitter in the initial distribution, the BW algo thinks itself to already be in an optimal state, updates the transtitions to reflect the actual distributions, and quits. **Thus, the NLTK BW is not broken**. Instead, a user must provide randomly distributed `_priors` and `_transition` matrices (i.e. values that are NOT equal but still sum to 1) by instantiating a `HiddenMarkovModelTagger` with the `_priors` and `_transitions` set to the randomized distributions you would like, and pass that tagger as the kwarg "model" to the `HiddenMarkovModelTrainer` (a really wonky, unclear input for something that should be required). 

I understand that the NLTK documentation is not meant to teach people how to perform HMMs, but the correct usage i've outlined above is not apparent from the docs (at least to me). I would really like to see some documentation to this effect in the docstring for `train_unsupervised`, or even the class definition of the `HiddenMarkovModelTrainer`. 

Going further, the default behavior of setting the prior and transition matrices to flat, equivalent distributions provides **NO** value. Either the `HiddenMarkovModelTrainer` should require an instantiated `HiddenMarkovModelTagger` object as an argument (the way Matlab does), or the default initialization of prior and transition matrices should be a randomized distribution. Frankly, the initial distribution for priors and transitions can be any random distribution (gaussian, uniform, etc.), as long as each element is not equal. I've written a basic unirand function below that could be used to initialize `_priors` and `_transitions`:

``` python
# Uniform random distribution creation for
# a single _state (need to do this for each _state)
from numpy import random

def unirand(M)
    randrow = random.random(M)
    randrow = randrow/sum(randrow)

    # Check that row equals 1, if not, subtract diff from last element
    if np.sum(randrow) != 1:
        randrow[-1] += 1-sum(randrow)

    return randrow
```

Because as is, `train_unsupervised` will never update itself without random inits (and it doesn't mention that to you). I'd be happy to submit a pull request with these changes added, if it seems like something that would be valuable. 

**tl;dr:**
-  `train_unsupervised` does not appear broken
-  flat initial distribuions for _priors and _transitions causes lack of algo update
-  We need to either A) make a `HiddenMarkovModelTagger` a required arg to the Trainer or B) change the default initialization of _priors and _transitions to the unirand function above.

**Comment by stevenbird** [comment]
Good idea, given that a uniform prior is evidently useless.

**Comment by kmike** [comment]
I think all issues described in this ticket are resolved.
